datachannel_js.go

```diff
@@ -0,0 +1,185 @@
+// +build js
+
+package webrtc
+
+import (
+	"syscall/js"
+)
+
+const dataChannelBufferSize = 16384 // Lowest common denominator among browsers
+
+// DataChannel represents a WebRTC DataChannel
+// The DataChannel interface represents a network channel
+// which can be used for bidirectional peer-to-peer transfers of arbitrary data
+type DataChannel struct {
+	// Pointer to the underlying JavaScript RTCPeerConnection object.
+	underlying js.Value
+
+	// Keep track of handlers/callbacks so we can call Release as required by the
+	// syscall/js API. Initially nil.
+	onOpenHandler    *js.Func
+	onCloseHandler   *js.Func
+	onMessageHandler *js.Func
+}
+
+// OnOpen sets an event handler which is invoked when
+// the underlying data transport has been established (or re-established).
+func (d *DataChannel) OnOpen(f func()) {
+	if d.onOpenHandler != nil {
+		oldHandler := d.onOpenHandler
+		defer oldHandler.Release()
+	}
+	onOpenHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go f()
+		return js.Undefined()
+	})
+	d.onOpenHandler = &onOpenHandler
+	d.underlying.Set("onopen", onOpenHandler)
+}
+
+// OnClose sets an event handler which is invoked when
+// the underlying data transport has been closed.
+func (d *DataChannel) OnClose(f func()) {
+	if d.onCloseHandler != nil {
+		oldHandler := d.onCloseHandler
+		defer oldHandler.Release()
+	}
+	onCloseHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go f()
+		return js.Undefined()
+	})
+	d.onCloseHandler = &onCloseHandler
+	d.underlying.Set("onclose", onCloseHandler)
+}
+
+// OnMessage sets an event handler which is invoked on a binary message arrival
+// from a remote peer. Note that browsers may place limitations on message size.
+func (d *DataChannel) OnMessage(f func(msg DataChannelMessage)) {
+	if d.onMessageHandler != nil {
+		oldHandler := d.onMessageHandler
+		defer oldHandler.Release()
+	}
+	onMessageHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		event := args[0]
+		rawData := event.Get("data").String()
+		msg := DataChannelMessage{
+			IsString: false,
+			Data:     []byte(rawData),
+		}
+		go f(msg)
+		return js.Undefined()
+	})
+	d.onMessageHandler = &onMessageHandler
+	d.underlying.Set("onmessage", onMessageHandler)
+}
+
+// Send sends the binary message to the DataChannel peer
+func (d *DataChannel) Send(data []byte) error {
+	return d.SendText(string(data))
```

Comment:
This seems wrong. There is actually a difference in sending a string vs a binary format in the data channel spec. I think this depends on the datatype passed to the method in the JS world. It would be nice if we could support that here as well.

Reply:
Hmm.. I've been using the MDN docs as a reference but they are light on the details of [what happens under the hood when you send different data types](https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/send). Is there another reference you have been using? How is the Go implementation able to distinguish between the different types of data?

Reply:
Actually never mind. I don't necessarily need to understand everything that's going on under the hood, just how to convert between the JavaScript types and Go representation of those types. This is sort of a huge can of worms and I'm not going to be able to support everything. I'm going to aim to support `USVString` <-> `string` and `ArrayBuffer` <-> `[]byte`, then return a reasonable error if we see a type we didn't expect.

Reply:
Ok, that sounds great. We can extend when we run into other types.

Reply:
Take a look at the latest commit. I think we are handling this more correctly now.